From 0ff6dee0dc6d198df999aee26c6add976712b42c Mon Sep 17 00:00:00 2001
From: Louis Leseur <louis.leseur@gmail.com>
Date: Thu, 21 Nov 2024 17:41:53 +0100
Subject: [PATCH] net/qed: allow old cards not supporting "num_images" to work

Commit 43645ce03e0063d7c4a5001215ca815188778881 added support for
populating flash image attributes, notably "num_images". However, some
cards were not able to return this information. In such cases, the
driver would return EINVAL, causing the driver to exit.

We added a check to return EOPNOTSUPP when the card is not able to
return these information, allowing the driver continue instead of
returning an error.

Co-developed-by: Florian Forestier <florian@forestier.re>
Signed-off-by: Florian Forestier <florian@forestier.re>
Signed-off-by: Louis Leseur <louis.leseur@gmail.com>
---
 drivers/net/ethernet/qlogic/qed/qed_mcp.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/qed_mcp.c b/qed_mcp.c
index 16e6bd466143..6218d9c26855 100644
--- a/qed_mcp.c
+++ b/qed_mcp.c
@@ -3314,7 +3314,9 @@ int qed_mcp_bist_nvm_get_num_images(struct qed_hwfn *p_hwfn,
 	if (rc)
 		return rc;
 
-	if (((rsp & FW_MSG_CODE_MASK) != FW_MSG_CODE_OK))
+	if (((rsp & FW_MSG_CODE_MASK) == FW_MSG_CODE_UNSUPPORTED))
+		rc = -EOPNOTSUPP;
+	else if (((rsp & FW_MSG_CODE_MASK) != FW_MSG_CODE_OK))
 		rc = -EINVAL;
 
 	return rc;
-- 
2.45.2

